File: JavaScript/boj/Gold/16987.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const [N, ...input] = fs.readFileSync(filePath).toString().trim().split('\n');

const eggs = input.map((egg) => egg.split(' ').map(Number));
let result = 0;

const dfs = (now) => {
  let pass = true;

  if (now === +N) {
    let count = 0;

    eggs.forEach(([d, w]) => {
      if (d <= 0) count++;
    });
    result = Math.max(result, count);

    return;
  }

  for (let i = 0; i < +N; i++) {
    if (eggs[now][0] <= 0 || eggs[i][0] <= 0 || i === now) continue;

    pass = false;
    eggs[now][0] -= eggs[i][1];
    eggs[i][0] -= eggs[now][1];
    dfs(now + 1);
    eggs[now][0] += eggs[i][1];
    eggs[i][0] += eggs[now][1];
  }

  if (pass) dfs(now + 1);
};

dfs(0);
console.log(result);

File: Python/boj/Gold/16987.py
import sys

if __name__ == "__main__":
    try:
        sys.stdin = open("input.txt", "r")
    except FileNotFoundError:
        pass

    N = int(input())
    eggs = [list(map(int, input().split())) for _ in range(N)]
    result = 0

    def dfs(now):
        global result
        hit = False

        if now == N:
            count = sum(1 for d, w in eggs if d <= 0)
            result = max(count, result)

            return

        for i in range(N):
            if i == now or eggs[i][0] <= 0 or eggs[now][0] <= 0:
                continue

            hit = True
            eggs[now][0] -= eggs[i][1]
            eggs[i][0] -= eggs[now][1]
            dfs(now + 1)
            eggs[now][0] += eggs[i][1]
            eggs[i][0] += eggs[now][1]

        if not hit:
            dfs(now + 1)

    dfs(0)
    print(result)
